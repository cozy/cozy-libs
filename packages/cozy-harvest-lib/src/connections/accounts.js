import merge from 'lodash/merge'

export const ACCOUNTS_DOCTYPE = 'io.cozy.accounts'
const PERMISSIONS_DOCTYPE = 'io.cozy.permissions'

/**
 * Create an account for the given konnector.
 * @param  {Object}  client     CozyClient
 * @param  {Object}  konnector  io.cozy.konnectors document
 * @param  {Object}  attributes Account attributes
 */
const createAccount = async (client, konnector, attributes) => {
  const { data } = konnector.aggregator
    ? await createChildAccount(client, konnector, attributes)
    : await client.create(ACCOUNTS_DOCTYPE, attributes)
  return data
}

/**
 * TODO Use get
 * Performs a query to find the given account
 * @param  {Object}  client CozyClient
 * @param  {string}  id     io.cozy.accounts document's id
 * @return {Object}         Retrieved account
 */
export const findAccount = async (client, id) => {
  try {
    const { data } = await client.query(
      client.find(ACCOUNTS_DOCTYPE).where({
        _id: id
      })
    )
    return (data && data[0]) || null
  } catch (error) {
    if (error.status === 404) {
      return null
    }
    throw error
  }
}

/**
 * Creates an account for an konnector needing an aggregator account.
 * Creates the aggregator account as well if it does not exist.
 * @param  {Object}  client     CozyClient
 * @param  {Object}  konnector  io.cozy.konnectors document
 * @param  {Object}  attributes Account attibutes
 * @return {Object}             The created io.cozy.accounts document
 */
const createChildAccount = async (client, konnector, attributes) => {
  const { aggregator } = konnector

  if (!aggregator || !aggregator.accountId)
    throw new Error('Konnector does not provide aggregator account id')

  const parentAccountId = aggregator.accountId

  let parentAccount
  try {
    parentAccount = await findAccount(client, parentAccountId)
  } catch (error) {
    throw new Error(
      `An error occurred when finding parent account ${parentAccountId} (${error.message})`
    )
  }

  if (!parentAccount) {
    try {
      parentAccount = await client.create(ACCOUNTS_DOCTYPE, {
        _id: parentAccountId
      })
    } catch (error) {
      throw new Error(
        `Cannot create parent account ${parentAccountId} (${error.message})`
      )
    }
  }

  try {
    await client.collection(PERMISSIONS_DOCTYPE).add(konnector, {
      aggregatorAccount: {
        type: ACCOUNTS_DOCTYPE,
        verbs: ['GET', 'PUT'],
        values: [`${ACCOUNTS_DOCTYPE}.${parentAccountId}`]
      }
    })
  } catch (error) {
    throw new Error(
      `Cannot set permission for account ${parentAccountId} (${error.message})`
    )
  }

  return await client.create(ACCOUNTS_DOCTYPE, {
    ...attributes,
    relationships: {
      parent: {
        data: {
          _id: parentAccountId,
          _type: ACCOUNTS_DOCTYPE
        }
      }
    }
  })
}

/**
 * Updates existing account
 * @param  {Object}  client  CozyClient
 * @param  {Object}  account io.cozy.accounts document to update
 * @return {Object}          Updated io.cozy.accounts document
 */
export const updateAccount = async (client, account) => {
  try {
    const { data } = await client.save(account)
    return data
  } catch (error) {
    if (error.status === 409) {
      /* We fetch before in case of conflicts since the account can be changed
      in the database by the konnector */
      const upToDateAccount = await findAccount(client, account._id)
      delete account._rev
      const { data } = await client.save(merge(upToDateAccount, account))
      return data
    } else {
      throw error
    }
  }
}

/**
 * Creates or updates account
 * @param  {Object}  client  CozyClient
 * @param  {Object}  konnector  io.cozy.konnectors document
 * @param  {Object}  authData   Account auth attribute
 */
const saveAccount = (client, konnector, account = {}) => {
  return account._id
    ? updateAccount(client, account)
    : createAccount(client, konnector, account)
}

/**
 * Deletes an account.
 * @param  {Object}  client  CozyClient
 * @param  {Object}  account io.cozy.accounts document
 */
export const deleteAccount = async (client, account) => {
  try {
    await client.destroy(account)
  } catch (error) {
    if (error.status === 409) {
      const syncedAccount = await findAccount(client, account._id)
      await client.destroy(syncedAccount)
    } else {
      throw error
    }
  }
}

/**
 * Get accounts mutations
 * @param  {Object} client CozyClient
 * @return {Object}        Object containing accounts mutations
 */
export const accountsMutations = client => ({
  createAccount: createAccount.bind(null, client),
  findAccount: findAccount.bind(null, client),
  updateAccount: updateAccount.bind(null, client),
  deleteAccount: deleteAccount.bind(null, client),
  saveAccount: saveAccount.bind(null, client)
})

export default accountsMutations
