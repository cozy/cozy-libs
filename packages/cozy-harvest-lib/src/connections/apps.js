import CozyClient, { Q } from 'cozy-client'

export const appsConn = {
  query: Q('io.cozy.apps'),
  as: 'harvest-apps',
  fetchPolicy: CozyClient.fetchPolicies.olderThan(30 * 1000)
}
