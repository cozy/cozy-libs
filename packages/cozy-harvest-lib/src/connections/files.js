const FILES_DOCTYPE = 'io.cozy.files'

/**
 * Add references to given files into given documents
 * @param  {Object}  client    CozyClient
 * @param  {Object}  document  Target document
 * @param  {[type]}  files     Files to reference
 * @return {Object}           Updated document
 */
export const addReferencesTo = async (client, document, files) => {
  const { data } = await client
    .collection(FILES_DOCTYPE)
    .addReferencesTo(document, files)
  return data
}

/**
 * Creates a directory from a given path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}         Directory attributes
 */
export const createDirectoryByPath = async (client, path) => {
  const { data } = await client
    .collection(FILES_DOCTYPE)
    .createDirectoryByPath(path)
  return data
}

/**
 * Retrieves a directory from its path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}        Created io.cozy.files document
 */
export const statDirectoryByPath = async (client, path) => {
  let response
  try {
    response = await client.collection(FILES_DOCTYPE).statByPath(path)
  } catch (error) {
    if (error && error.status === 404) return null
    throw new Error(error.message)
  }

  return response.data
}

export const filesMutations = client => ({
  addReferencesTo: addReferencesTo.bind(null, client),
  createDirectoryByPath: createDirectoryByPath.bind(null, client),
  statDirectoryByPath: statDirectoryByPath.bind(null, client)
})

export default filesMutations
