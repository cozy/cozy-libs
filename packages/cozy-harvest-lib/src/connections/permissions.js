const PERMISSIONS_DOCTYPE = 'io.cozy.permissions'

/**
 * Add the given permisison to the given subject
 * @param {Object} client     CozyClient
 * @param {Object} target     Target document, could have doctype io.cozy.apps,
 * io.cozy.konnectors or io.cozy.permissions
 * @param {{ type, values, verbs}} A permission objects
 */
const addPermission = async (client, target, permission) => {
  const { data } = await client
    .collection(PERMISSIONS_DOCTYPE)
    .add(target, permission)
  return data
}

export const permissionsMutations = client => ({
  addPermission: addPermission.bind(null, client)
})

export default permissionsMutations
