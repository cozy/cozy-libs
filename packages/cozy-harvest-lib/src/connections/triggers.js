import KonnectorJobWatcher from '../models/konnector/KonnectorJobWatcher'

const TRIGGERS_DOCTYPE = 'io.cozy.triggers'

/**
 * Creates a trigger with given attributes
 * @param  {Object} client CozyClient
 * @param  {Object}   attributes
 * @return {Object}   Created trigger
 */
const createTrigger = async (client, attributes) => {
  const { data } = await client.collection(TRIGGERS_DOCTYPE).create(attributes)
  return data
}

const fetchTrigger = async (client, id) => {
  const { data } = await client.collection(TRIGGERS_DOCTYPE).get(id)
  return data
}

/**
 * Triggers job associated to given trigger
 * @param  {Object} client CozyClient
 * @param  {Object}  Trigger to launch
 * @return {Object}  Job document
 */
const launchTrigger = async (client, trigger) => {
  const { data } = await client.collection(TRIGGERS_DOCTYPE).launch(trigger)
  return data
}

/**
 * This method is here to listen the jobs.
 *
 * expectedSuccessDelay: Is the time (in ms) before dispatching
 * a `LOGIN_SUCESS` event if nothing was received before this delay
 * (error, success or login success)
 *
 * It means that if the login process takes more than this time, we
 * will have a `LOGIN_SUCESS` even if it's not the case.
 *
 * @param {Object} client CozyClient
 * @param {Object} job
 *
 */
const watchKonnectorJob = (client, job) => {
  const jobWatcher = new KonnectorJobWatcher(client, job, {
    expectedSuccessDelay: 8000
  })
  // no need to await realtime initializing here
  jobWatcher.watch()
  return jobWatcher
}

/**
 * Return triggers mutations
 * @param  {Object} client CozyClient
 * @return {Object}        Object containing mutations
 */
export const triggersMutations = client => {
  return {
    createTrigger: createTrigger.bind(null, client),
    fetchTrigger: fetchTrigger.bind(null, client),
    launchTrigger: launchTrigger.bind(null, client),
    watchKonnectorJob: watchKonnectorJob.bind(null, client)
  }
}

export default triggersMutations
