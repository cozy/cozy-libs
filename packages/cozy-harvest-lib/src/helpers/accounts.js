import get from 'lodash/get'
import merge from 'lodash/merge'

import manifest from './manifest'
import { saveAccount as saveAccountMutation } from '../connections/accounts'
const DEFAULT_TWOFA_CODE_PROVIDER_TYPE = 'default'

const PROVIDERS = {
  EMAIL: 'email',
  SMS: 'sms'
}

const TWOFA_NEEDED_STATUS = 'TWOFA_NEEDED'
const TWOFA_NEEDED_RETRY_STATUS = 'TWOFA_NEEDED_RETRY'
const RESET_SESSION_STATE = 'RESET_SESSION'
const HANDLE_LOGIN_SUCCESS_STATE = 'HANDLE_LOGIN_SUCCESS'
const LOGIN_SUCCESS_STATE = 'LOGIN_SUCCESS'

/**
 * Return a boolean to know if the account is in a two fa code needed
 * status
 * @param  {String}  status Account two FA Status
 * @return {Boolean}
 */
export const isTwoFANeeded = status => {
  if (!status) return false
  return status.split('.')[0] === TWOFA_NEEDED_STATUS
}

export const isTwoFARetry = status => {
  if (!status) return false
  return status.split('.')[0] === TWOFA_NEEDED_RETRY_STATUS
}

export const isLoginSuccessHandled = status => {
  return status === HANDLE_LOGIN_SUCCESS_STATE
}

export const isLoginSuccess = status => {
  return status === LOGIN_SUCCESS_STATE
}

/**
 * Return the status object key matching the status value
 * @param  {String} status Account document
 * @return {String}        Two FA Code providing type or default one if not known
 */
export const getTwoFACodeProvider = account => {
  if (!account || !account.state) return DEFAULT_TWOFA_CODE_PROVIDER_TYPE
  const codeParts = account.state ? account.state.split('.') : []
  if (codeParts.length > 1) {
    return PROVIDERS[codeParts[1]] || DEFAULT_TWOFA_CODE_PROVIDER_TYPE
  } else {
    return DEFAULT_TWOFA_CODE_PROVIDER_TYPE
  }
}

/**
 * Returns the label for the given account.
 * This label is by default the value for the identifier field.
 * If there is no value for this field, the label is the io.cozy.accounts
 * document id.
 * @param  {Object} account io.cozy.accounts documents
 * @return {string}         The label associated to this account.
 */
export const getLabel = account =>
  get(account, `auth.${account.identifier}`) || account._id

/**
 * Transforms AccountForm data to io.cozy.accounts document
 * @param  {object} konnector Konnector related to account
 * @param  {object} data      Data from AccountForm
 * @return {object}           io.cozy.accounts attributes
 */
export const build = (konnector, authData) => {
  // We are not at the final target for io.cozy.accounts.
  // For now we are just ensuring legacy
  return {
    auth: authData,
    account_type: konnector.slug,
    identifier: manifest.getIdentifier(konnector.fields),
    state: null
  }
}

/**
 * Merges existing io.cozy.accounts auth with Auth data from AccountForm
 * @param  {object} account   io.cozy.accounts document
 * @param  {object} data      Data from AccountForm
 * @return {object}           io.cozy.accounts attributes
 */
export const mergeAuth = (account, authData) => ({
  ...account,
  auth: merge(account.auth, authData)
})

/**
 * Update Two FA code from TwoFAForm into io.cozy.accounts document
 * @param  {object} account   io.cozy.accounts document
 * @param  {object} code      Code from TwoFAForm
 * @return {object}           io.cozy.accounts attributes
 */
export const updateTwoFaCode = (account, code) => ({
  // reset the state since the konnector is listening it
  ...resetState(account),
  twoFACode: code
})

/**
 * Reset the account state
 * @param  {Object} account Account document
 * @return {object}         Changed account document
 */
export const resetState = account => ({
  ...account,
  state: null
})

/**
 * Set a state to reset the konnector session into io.cozy.accounts document
 * only if necessary, if password/passphrase have changed
 * @param  {object} account   io.cozy.accounts document
 * @return {object}           io.cozy.accounts updated document
 */
export const setSessionResetIfNecessary = (account, changedFields) => {
  const isPasswordChanged =
    !!account && !!(changedFields.password || changedFields.passphrase)
  return isPasswordChanged
    ? {
        ...account,
        state: RESET_SESSION_STATE
      }
    : account
}

/**
 * Update an account for a given konnector
 * TODO: Move to cozy-doctypes
 */
export const updateAccount = async ({ konnector, account, data, client }) => {
  const mergedAccount = mergeAuth(
    setSessionResetIfNecessary(resetState(account), data),
    data
  )
  return saveAccount(mergedAccount, data, client, konnector)
}

/**
 * Create an account for a given konnector
 * TODO: Move to cozy-doctypes
 */
export const createAccount = async ({ konnector, data, client }) => {
  const accountDocument = build(konnector, data)
  return saveAccount(accountDocument, data, client, konnector)
}

/**
 * Save the account and return a mergeAuth account
 * TODO: Move to cozy-doctypes
 */
const saveAccount = async (account, data, client, konnector) => {
  const savedAccount = mergeAuth(
    await saveAccountMutation(client, konnector, account),
    data
  )
  return savedAccount
}

export default {
  build,
  getLabel,
  getTwoFACodeProvider,
  isTwoFANeeded,
  isTwoFARetry,
  isLoginSuccess,
  isLoginSuccessHandled,
  mergeAuth,
  resetState,
  setSessionResetIfNecessary,
  updateTwoFaCode,
  updateAccount,
  createAccount
}
