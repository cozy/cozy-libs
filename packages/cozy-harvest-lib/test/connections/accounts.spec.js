/* eslint-env jest */

import { accountsMutations } from 'connections/accounts'

const client = {
  collection: jest.fn().mockReturnValue({
    add: jest.fn(),
    get: jest.fn()
  }),
  create: jest.fn(),
  destroy: jest.fn(),
  query: jest.fn(),
  save: jest.fn(),
  where: jest.fn(),
  stackClient: {
    uri: 'cozy.tools:8080',
    token: {
      token: '1234abcd'
    }
  },
  find: jest.fn().mockReturnValue({
    where: jest.fn()
  }),
  options: {
    uri: 'cozy.tools:8080'
  },
  on: () => jest.fn()
}

const fixtures = {
  accountPermissions: {
    source_id: `io.cozy.konnectors/kaggregated`,
    permissions: {
      aggregatorAccount: {
        type: 'io.cozy.accounts',
        verbs: ['GET', 'PUT'],
        values: ['io.cozy.accounts.kaggregated-aggregator']
      }
    }
  },
  konnector: {
    slug: 'konnectest'
  },
  konnectorWithAggregator: {
    slug: 'kaggregated',
    aggregator: {
      accountId: 'kaggregated-aggregator'
    }
  },
  simpleAccount: {
    account_type: 'test',
    auth: {
      login: 'login',
      password: 'pass'
    }
  },
  existingAccount: {
    _id: '561be660ff384ce0846c8f20e829ad62',
    account_type: 'test',
    auth: {
      login: 'login',
      password: 'pass'
    }
  },
  parentAccount: {
    _id: 'kaggregated-aggregator'
  }
}

const {
  createAccount,
  updateAccount,
  saveAccount,
  deleteAccount,
  watchKonnectorAccount
} = accountsMutations(client)

describe('Account mutations', () => {
  beforeEach(() => {
    client.create.mockResolvedValue({ data: fixtures.existingAccount })
    client.save.mockResolvedValue({ data: fixtures.existingAccount })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  describe('createAccount', () => {
    it('calls Cozy Client and return account', async () => {
      const account = await createAccount(
        fixtures.konnector,
        fixtures.simpleAccount
      )

      expect(client.create).toHaveBeenCalledWith(
        'io.cozy.accounts',
        fixtures.simpleAccount
      )
      expect(account).toEqual(fixtures.existingAccount)
    })

    const simpleAccountFixtureWithMasterRelation = {
      ...fixtures.simpleAccount,
      relationships: {
        parent: {
          data: {
            _id: 'kaggregated-aggregator',
            _type: 'io.cozy.accounts'
          }
        }
      }
    }

    beforeEach(() => {
      client.query.mockResolvedValue({
        data: simpleAccountFixtureWithMasterRelation
      })
    })

    it('throws an error when konnector have no aggregator.accountId attribute', async () => {
      const konnector = { aggregator: {} }
      await expect(
        createAccount(konnector, fixtures.simpleAccount)
      ).rejects.toEqual(
        new Error('Konnector does not provide aggregator account id')
      )
    })

    describe('if parent account does not exists', () => {
      beforeEach(() => {
        client.collection().get = jest.fn().mockRejectedValue({ status: 404 })

        client.create.mockReset()
        client.create.mockResolvedValue({
          data: simpleAccountFixtureWithMasterRelation
        })
        client.create.mockResolvedValueOnce({ data: fixtures.parentAccount })
      })

      it('creates parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith('io.cozy.accounts', {
          _id: fixtures.konnectorWithAggregator.aggregator.accountId
        })
      })

      it('adds permissions to parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.collection().add).toHaveBeenCalledWith(
          fixtures.konnectorWithAggregator,
          fixtures.accountPermissions.permissions
        )
      })

      it('creates child account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith(
          'io.cozy.accounts',
          simpleAccountFixtureWithMasterRelation
        )
      })

      it('throws error if get fails', async () => {
        client.query.mockReset()
        client.query.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'An error occurred when finding parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if permission add fails', async () => {
        client
          .collection()
          .add.mockRejectedValueOnce(new Error('Mocked add Error'))
        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'Cannot set permission for account kaggregated-aggregator (Mocked add Error)'
          )
        )
      })

      it('returns child account', async () => {
        const account = await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(account).toEqual(simpleAccountFixtureWithMasterRelation)
      })

      it('throws error if parent account creation fails', async () => {
        client.create.mockReset()
        client.create.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'Cannot create parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if child account creation fails', async () => {
        client.create.mockReset()
        client.create.mockResolvedValueOnce({ data: fixtures.parentAccount })
        client.create.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(new Error('Mocked error'))
      })
    })

    describe('if parent account exists', () => {
      beforeEach(() => {
        client.query.mockReset()
        client.query = jest
          .fn()
          .mockResolvedValue({ data: [fixtures.parentAccount] })
        client.create.mockResolvedValue({
          data: simpleAccountFixtureWithMasterRelation
        })
      })

      it('does not create parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).not.toHaveBeenCalledWith('io.cozy.accounts', {
          _id: fixtures.konnectorWithAggregator.aggregator.accountId
        })
      })

      it('adds permissions to parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.collection().add).toHaveBeenCalledWith(
          fixtures.konnectorWithAggregator,
          fixtures.accountPermissions.permissions
        )
      })

      it('creates child account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith(
          'io.cozy.accounts',
          simpleAccountFixtureWithMasterRelation
        )
      })

      it('returns child account', async () => {
        const account = await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(account).toEqual(simpleAccountFixtureWithMasterRelation)
      })

      it('throws error if get fails', async () => {
        client.query.mockReset()
        client.query.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'An error occurred when finding parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if child account creation fails', async () => {
        client.create.mockReset()
        client.create.mockRejectedValueOnce(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(new Error('Mocked error'))
      })
    })
  })

  describe('updateAccount', () => {
    it('calls CozyClient::save and returns account', async () => {
      const account = await updateAccount(fixtures.simpleAccount)
      expect(client.save).toHaveBeenCalledWith(fixtures.simpleAccount)
      expect(account).toEqual(fixtures.existingAccount)
    })
  })

  describe('deleteAccount', () => {
    it('calls CozyClient::destroy', async () => {
      await deleteAccount(fixtures.simpleAccount)
      expect(client.destroy).toHaveBeenCalledWith(fixtures.simpleAccount)
    })
  })

  describe('saveAccount', () => {
    it('calls CozyClient::create for new account', async () => {
      const account = await saveAccount(
        fixtures.konnector,
        fixtures.simpleAccount
      )

      expect(client.create).toHaveBeenCalledWith(
        'io.cozy.accounts',
        fixtures.simpleAccount
      )
      expect(account).toEqual(fixtures.existingAccount)
    })

    it('calls CozyClient::save for existing account', async () => {
      const account = await saveAccount(
        fixtures.konnector,
        fixtures.existingAccount
      )
      expect(client.save).toHaveBeenCalledWith(fixtures.existingAccount)
      expect(account).toEqual(fixtures.existingAccount)
    })
  })

  describe('watchAccount', () => {
    it('should return an account watcher with correct properties', async () => {
      const mockAccount = {
        accountType: 'dummy',
        auth: {}
      }
      const accountWatcher = await watchKonnectorAccount(mockAccount, {
        onError: jest.fn(),
        onLoginSuccess: jest.fn(),
        onSuccess: jest.fn()
      })
      expect(accountWatcher).toMatchSnapshot()
    })
  })
})
