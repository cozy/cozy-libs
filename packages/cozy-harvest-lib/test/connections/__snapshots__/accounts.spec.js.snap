// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account mutations watchAccount should return an account watcher with correct properties 1`] = `
KonnectorAccountWatcher {
  "account": Object {
    "accountType": "dummy",
    "auth": Object {},
  },
  "handleAccountUpdated": [Function],
  "options": Object {
    "onError": [MockFunction],
    "onLoginSuccess": [MockFunction],
    "onSuccess": [MockFunction],
  },
  "realtime": CozyRealtime {
    "_beforeUnload": [Function],
    "_cozyClient": Object {
      "collection": [MockFunction],
      "create": [MockFunction],
      "destroy": [MockFunction],
      "find": [MockFunction],
      "on": [Function],
      "options": Object {
        "uri": "cozy.tools:8080",
      },
      "query": [MockFunction],
      "save": [MockFunction],
      "stackClient": Object {
        "token": Object {
          "token": "1234abcd",
        },
        "uri": "cozy.tools:8080",
      },
      "where": [MockFunction],
    },
    "_events": Object {
      "updated//io.cozy.accounts//undefined": Array [
        [Function],
      ],
    },
    "_receiveError": [Function],
    "_receiveMessage": [Function],
    "_resetSocket": [Function],
    "_resubscribe": [Function],
    "_retryDelay": 1000,
    "_retryLimit": 60,
    "_socket": Socket {
      "_doctype": "io.cozy.websocket",
      "_events": Object {
        "close": Array [
          [Function],
        ],
        "error": Array [
          [Function],
        ],
        "message": Array [
          [Function],
        ],
        "subscribe_io.cozy.accounts_undefined": Array [
          [Function],
        ],
      },
      "_getToken": [Function],
      "_getUrl": [Function],
      "_token": null,
      "_url": null,
      "_webSocket": WebSocket {},
      "removeAllListeners": [Function],
    },
    "_updateAuthentication": [Function],
    "unsubscribeAll": [Function],
  },
}
`;
